lib/src/merge.dart

```diff
@@ -63,47 +63,46 @@ class _Merge<T> extends StreamTransformerBase<T, T> {
 
   @override
   Stream<T> bind(Stream<T> first) {
-    var controller = first.isBroadcast
+    final controller = first.isBroadcast
         ? StreamController<T>.broadcast(sync: true)
         : StreamController<T>(sync: true);
 
-    var allStreams = [first]..addAll(_others);
-    if (first.isBroadcast) {
-      allStreams = allStreams
-          .map((s) => s.isBroadcast ? s : s.asBroadcastStream())
-          .toList();
-    }
-
-    List<StreamSubscription<T>> subscriptions;
+    final allStreams = [
+      first,
+      for (final other in _others)
+        !first.isBroadcast || other.isBroadcast
+            ? other
+            : other.asBroadcastStream(),
+    ];
 
     controller.onListen = () {
-      assert(subscriptions == null);
-      var activeStreamCount = 0;
-      subscriptions = allStreams.map((stream) {
-        activeStreamCount++;
-        return stream.listen(controller.add, onError: controller.addError,
-            onDone: () {
-          if (--activeStreamCount <= 0) controller.close();
+      final subscriptions = <StreamSubscription<T>>[];
+      for (final stream in allStreams) {
+        final subscription =
+            stream.listen(controller.add, onError: controller.addError);
+        subscription.onDone(() {
+          assert(subscriptions.contains(subscription));
```

Comment:
Why is the assert necessary here and the similar one above?

Reply:
Taking a look again, I think I had at one point during the refactor had a setup where the `subscriptions.remove` could happen in multiple places. Since this is now the only place that happens, and we know a stream will only end once, I agree that it's trivially safe. I'll remove them.